ecl/eclagent/eclagent.cpp

```diff
@@ -1950,6 +1951,25 @@ void EclAgent::doProcess()
             break;
         }
 
+        if (w->getState() == WUStateCompleted && getClusterType(clusterType)==ThorLCRCluster)
+        {
+            if (agentTopology->getPropBool("@analyzeWorkunit", true) && w->getDebugValueBool("analyzeWorkunit", true))
+            {
+                WuAnalyseOptions options;
+                CIArrayOf<PerformanceIssue> issues;
+                analyseWorkunit(w.get(), options, issues);
+                ForEachItemIn(i, issues)
```

Comment:
I preferred the structure of the previous commit.  You are breaking encapsulation of the analysis logic - previously these classes were not part of the public interface (and they are not exported from the dll).  It also means you have lost the option to run the analyser from the wutool and write the results to the workunit.  I can imagine that being useful - e.g., if you want to run new rules on an old workunit.

Reply:
My thinking was that analyseWorkunit would do only that - analyse.  Updating the workunit or printing out the results doesn't seem part of the analysis. We could provide a separate function to write the issues to the WU and another function to print out the issues.

The other question is that if we were to do it the way it was done before, would we pass a locked or an unlocked workunit?  A locked workunit is not necessary for Analyze & Print but an locked workunit is required for Analyse & Update WU.  It seems that we'd need to either lock WU unnessary or need to relock the WU (that may already be locked).

Reply:
There is always a balancing act for where code goes - but generally it is best to try and encapsulate code to minimize information leaking.  For instance if the results of the analysis were then going to be processed by multiple other stages then it would make sense to export the information about the results.  If (as in this case) the results have little further processing it makes sense to keep the details hidden.

As for the details, in this case I would have had two functions, one that took a IWorkUnit and added the results to the workunit exceptions, the other than took an IConstWorkUnit (which is a base interface of IWorkUnit) and printed them.  They would use common code within the anawu module.

Slightly related one rule I have found useful is that if most of the code within a function is accessing members of an object that as a good indication that should be a member function.  Which would argue against removing PerformanceIssue::print() - and similarly the code to add an issue as a workunit exception would be cleaner as a member function.